service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -31,153 +26,57 @@
-    private final OrganizationToAgentMapper organizationToAgentMapper;
+    private final MessageContext messageContext;
+    private final AgentPersonMapper agentPersonMapper;
 
     public String mapEHRFolderToAgentDirectory(Bundle bundle, String nhsNumber) {
-        Set<String> mappedOrganizationsAndPractitioner = new HashSet<>();
-        var builder = AgentDirectoryParameter.builder()
-            .patientManagingOrganization(extractPatientManagingOrganization(bundle, nhsNumber))
-            .observationOrganizations(prepareObservationAgentsList(bundle, mappedOrganizationsAndPractitioner))
-            .referralRequestsOrganizations(prepareReferralRequestAgentsList(bundle, mappedOrganizationsAndPractitioner))
-            .agentPersons(preparePractitionerRoleAgentsList(bundle, mappedOrganizationsAndPractitioner));
+        var builder = AgentDirectoryParameter.builder();
 
-        return TemplateUtils.fillTemplate(AGENT_DIRECTORY_STRUCTURE_TEMPLATE, builder.build());
-    }
+        Set<Map.Entry<AgentDirectory.AgentKey, String>> agentDirectories = messageContext.getAgentDirectory().getEntries();
 
-    private String extractPatientManagingOrganization(Bundle bundle, String nhsNumber) {
-        return AgentDirectoryExtractor.extractPatientByNhsNumber(bundle, nhsNumber)
-            .filter(Patient::hasManagingOrganization)
-            .map(patient -> ResourceExtractor.extractResourceByReference(bundle,
-                patient.getManagingOrganization().getReferenceElement()))
-            .map(Optional::get)
-            .map(Organization.class::cast)
-            .map(organizationToAgentMapper::mapOrganizationToAgent)
+        Optional<Organization> patientManagingOrganization = extractPatientManagingOrganization(bundle, nhsNumber);
+
+        patientManagingOrganization
+            .map(organization -> OrganizationToAgentMapper.mapOrganizationToAgent(organization, getAgentId(organization)))
+            .map(builder::patientManagingOrganization)
             .orElseThrow(() -> new EhrMapperException("The ASR bundle does not contain a Patient resource with the correct identifier "
                 + "and a managingOrganization"));
-    }
-
-    private List<String> prepareObservationAgentsList(Bundle bundle, Set<String> mappedOrganizationsAndPractitioner) {
-        List<Observation> observations = AgentDirectoryExtractor.extractObservationsWithPerformers(bundle);
-
-        return observations.stream()
-            .map(observation -> buildAgentPerson(observation, bundle, mappedOrganizationsAndPractitioner))
-            .filter(StringUtils::isNotEmpty)
-            .collect(Collectors.toList());
-    }
 
-    private List<String> prepareReferralRequestAgentsList(Bundle bundle, Set<String> mappedOrganizationsAndPractitioner) {
-        List<ReferralRequest> referralRequests = AgentDirectoryExtractor.extractReferralRequestsWithRequester(bundle);
+        builder.agentPersons(
+            agentDirectories.stream()
```

Comment:
I don't understand this code. Could we make it a bit more declarative?
Why is agentDirectories being filtered? I expect this to be a set of every agent that goes into the directory. Doesn't adding a filter step mean we're going to be missing agents that were referenced by our previously output statements?

Reply:
The ides behind this is:
- first agent on agent directory is always with agentOrganization from patient.managingOrganization ( it is described in mapping excel, https://kainossoftwareltd-my.sharepoint.com/:x:/g/personal/davidmcav_kainos_com/Ef-K0P2Do9tMu13fI1gLkDsBES6fhhj8NN3pYzG_F3iJ5w?e=iiTKx9 on tab agentDirectory, at the top, also in ac of NIAd-1024
- then all other references go, so if other resource is referencing that organization also ( for example, observation.performer is sometimes referencing only organization) and in that case I want to map organization only once, so if it was mapped for agentOrganization first with the transformed id from AgenDirectory class, then it should not be mapped to normal agent again. 

But it is only applied when there is a single Organization, for Practitioner + Organization it will create agent with agentPerson and representedOrganization still.

What do you think about it now?

Reply:
Yes that makes sense now, and I see there is a test that describes this behaviour also When_MappingAgentDirectoryWithPatientManagingOrganizationInAgentKeys_Expect_AgentPersonNotDuplicated. I should have looked a bit deeper. :thumbsup: